// react-simple-di/dist/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.injectDeps = injectDeps;
exports.useDeps = useDeps;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _hoistNonReactStatics = require('hoist-non-react-statics');

var _hoistNonReactStatics2 = _interopRequireDefault(_hoistNonReactStatics);

var _reactDisplayName = require('react-display-name');

var _reactDisplayName2 = _interopRequireDefault(_reactDisplayName);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function injectDeps(context, _actions) {
  var actions = {};
  for (var key in _actions) {
    if (_actions.hasOwnProperty(key)) {
      var actionMap = _actions[key];
      var newActionMap = {};
      for (var actionName in actionMap) {
        if (actionMap.hasOwnProperty(actionName)) {
          newActionMap[actionName] = actionMap[actionName].bind(null, context);
        }
      }
      actions[key] = newActionMap;
    }
  }

  return function (Component) {
    var ComponentWithDeps = _react2.default.createClass({
      displayName: 'ComponentWithDeps',

      childContextTypes: {
        context: _react2.default.PropTypes.object,
        actions: _react2.default.PropTypes.object
      },

      getChildContext: function getChildContext() {
        return {
          context: context,
          actions: actions
        };
      },
      render: function render() {
        return _react2.default.createElement(Component, this.props);
      }
    });

    ComponentWithDeps.displayName = 'WithDeps(' + (0, _reactDisplayName2.default)(Component) + ')';
    return (0, _hoistNonReactStatics2.default)(ComponentWithDeps, Component);
  };
}

var defaultMapper = function defaultMapper(_context, _actions2) {
  return {
    context: function context() {
      return _context;
    },
    actions: function actions() {
      return _actions2;
    }
  };
};

function useDeps() {
  var mapper = arguments.length <= 0 || arguments[0] === undefined ? defaultMapper : arguments[0];

  return function (Component) {
    var ComponentUseDeps = _react2.default.createClass({
      displayName: 'ComponentUseDeps',
      render: function render() {
        var _context2 = this.context;
        var context = _context2.context;
        var actions = _context2.actions;

        var mappedProps = mapper(context, actions);

        var newProps = (0, _extends3.default)({}, this.props, mappedProps);

        return _react2.default.createElement(Component, newProps);
      },

      contextTypes: {
        context: _react2.default.PropTypes.object,
        actions: _react2.default.PropTypes.object
      }
    });

    ComponentUseDeps.displayName = 'UseDeps(' + (0, _reactDisplayName2.default)(Component) + ')';
    return (0, _hoistNonReactStatics2.default)(ComponentUseDeps, Component);
  };
}
